### PR TITLE
[expo-updates] Fix fingerprint runtime version policy

### PR DESCRIPTION
**Pull request by wschurman**
# Why

Alternative to https://github.com/expo/expo/pull/26839 that runs the fingerprint step at build time (same time as manifest generation happens). That PR summary should be read to see the context behind this PR.

This PR works better since it doesn't suffer from config plugin ordering concerns or arbitrary platform ordering. But it still suffers from the issue of most builders only creating a single platform, while updates are generally created on a machine that has both platforms set up (or sometimes neither pre-built). This is solved by doing per-platform fingerprinting, which will require updates to be also created in a per-platform manner so that the fingerprint runs only for that platform to match the fingerprint in the build. The second concern (sometimes updates are created on machines that haven't done pre-build) is addressed by detecting the workflow (generic or managed/CNG) using the same mechanism we do in eas-cli.
1. when using generic - hashes/fingerprints each platform (including the platform native root dir) separately at build time. For updates, a change in eas-cli will ensure that the updates being created are per-platform and that prompt the user to confirm that the native directory for that platform is in a good state before the runtime version is evaluated.
2. managed/CNG - hashes everything except the native root dirs. this doesn't need to be done per-platform since the fingerprint should be consistent no matter whether prebuild has been run and no matter whether it is being run as part of the android or ios build process or creation of an update.

The corresponding eas-cli draft PR for this proposal is https://github.com/expo/eas-cli/pull/2225

# How

When the fingerprint policy is used:
1. Set the runtime version in the native config files to `file:fingerprint`
2. At build time, create a file called `fingerprint` to the native bundle/assets containing the fingerprint for that platform.
4. When the configuration is loaded, load the runtime version from the fingerprint file.

# Test Plan

```
et gba -n testwat47 expo-file-system expo-font expo-modules-core  expo-updates expo-updates-interface expo-constants expo-keep-awake
cd bare-apps/testwat47
eas init
eas update:configure
// change runtime version in app.json to `fingerprintExperimental`
npx expo prebuild
cd ios
pod install
// run release builds on both android and ios and ensure the runtime version is correctly read
```

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by wschurman** [comment]
So now this theoretically works per-platform, so what's left is figuring out how to get this same per-platform fingerprint at time of running `eas update`. Some ideas:
- Require the update to be created from the same build machine. This doesn't work since we don't want to run builds on every commit, whereas we do want to run `eas update` on every commit. We only want to run builds when the fingerprint changes, so we need to be able to either detect that or preempt the build step in the eas builder to only do a `eas update`.
- Require pre-build from the triggering machine (gha runner, local) before deciding whether to run `eas update`/`eas build`. This is possibly more correct but then requires the triggering machine to be mac for ios. At this point it might even be better to just do it on eas somehow.

File: packages/@expo/config-types/src/ExpoConfig.ts
Comment (wschurman) [inline packages/@expo/config-types/src/ExpoConfig.ts:3]: These changes need to be made in www xdl-schemas, I just did them here to ease review. Will revert and change upstream and import before landing this.

**Comment by Kudo** [comment]
> The second concern (sometimes updates are created on machines that haven't done pre-build) is addressed by introducing two separate policies:

running fingerprint at build time is super clever. will take a detailed review tomorrow. before that i am thinking if we could solve the issue fundamentally without introducing two policies. some random idea:
could we pull some current state into runtime version? e.g. `runtimeVersion={hasPrebuild}:{fingerprint}`. when `hasPrebuild` is false and people running `eas update` after prebuild, we could show an error message. if we can restrict running eas update in same prebuild state, we may able to narrow down the edge cases.

**Comment by wschurman** [comment]
> > The second concern (sometimes updates are created on machines that haven't done pre-build) is addressed by introducing two separate policies:
> 
> running fingerprint at build time is super clever. will take a detailed review tomorrow. before that i am thinking if we could solve the issue fundamentally without introducing two policies. some random idea: could we pull some current state into runtime version? e.g. `runtimeVersion={hasPrebuild}:{fingerprint}`. when `hasPrebuild` is false and people running `eas update` after prebuild, we could show an error message. if we can restrict running eas update in same prebuild state, we may able to narrow down the edge cases.

Yeah, this is the area that is tough. expo-updates supports bare with prebuild run at every commit, bare with prebuild run once at upgrade and then ios/android dirs checked in, bare with prebuild run once a long time ago and manual changes applied to android/ios dirs, bare without prebuild, and managed (which uses prebuild during build?). So this limits our ability to "automatically" do stuff for users. In fact, it might even make sense to have three policies:
- `fingerprintNativeWithContinuousPrebuildExperimental` - in this case we could run prebuild for the user during `eas update` automatically
- `fingerprintNativeUserManagedExperimental` - in this case the user assumes responsibility for the state of their native project at time of fingerprint, either in build or in `eas update`.
- `fingerprintManagedExperimental` - in this case we only fingerprint the non-native stuff

I think it might be impossible to tell automatically whether someone a) has ever used prebuild b) still uses prebuild continuously, or c) wants to continue to use prebuild. We could theoretically add some sort of dry-run flag to prebuild to see if it would change the state or something.

File: packages/expo-updates/scripts/createFingerprint.js
Comment (quinlanj) [inline packages/expo-updates/scripts/createFingerprint.js:52]: IIRC the fingerprint package has a `reasons` JSON output to explain how the hash was created (it lists all the dependencies it creates the hash over). People will want to publish an update, targeting a build in the appstore. If the update fingerprint is different from the build fingerprint, we should make it easy for them to see why the fingerprint was different. We should consider saving the `reasons` JSON too

**Review by quinlanj — APPROVED** [review]
Code lgtm, but will defer to doug and kudo who is more familiar with it.

Couple comments:
- I'd prefer to keep the fingerprint policy as `fingerprintExperimental`, since we can automatically determine whether someone has a managed or bare workflow. To add onto what Kudo suggested, maybe something like `${appVersion}:${bareOrManaged}:${fingerprint}`, (ie) `1.0.0:bare:a1b2c3`
- For the first version of our experimental fingerprint, we can require the user have an osx machine to run `prebuild` if they are using the `bare` workflow. This can buy us time to figure out how we can set up a service to run fingerprint for non-osx folks (can leverage the custom jobs stuff that stanley is working on)

**Comment by wschurman** [comment]
> since we can automatically determine whether someone has a managed or bare workflow

ooh, how?

**Comment by wschurman** [comment]
> * For the first version of our experimental fingerprint, we can require the user have an osx machine to run `prebuild` if they are using the `bare` workflow. This can buy us time to figure out how we can set up a service to run fingerprint for non-osx folks (can leverage the custom jobs stuff that stanley is working on)

I think the issue of ios vs android was solved by just fingerprinting the platforms separately and creating updates separately as well, so we don't have to worry about running ios prebuild on a non-mac.

> * To add onto what Kudo suggested, maybe something like `${appVersion}:${bareOrManaged}:${fingerprint}`

Not sure I follow about the addition of `appVersion` here. Is this to give the user more control over implicit breaking changes?

**Comment by wschurman** [comment]
After discussing with @brentvatne on slack, here's the planned changes:

- one policy, `fingerprintExperimental` (same as current state of the world)
- we look and see whether ios/android are gitignored (determining CNG/managed vs bare)
- if they are ignored, we exclude ios/android from fingerprint hash
- if they are not ignored, we include them in the hash and warn the user that they are responsible for ensuring their native stuff is ready to be hashed (since we can’t tell if they’re using prebuild)

**Comment by Kudo** [comment]
i like the direction of having just one policy as the first version. even though expo-updates supports different workflows, we could limit the team running `eas update` from the same workflow. that is exactly what `runtimeVersion={hasPrebuild}:{fingerprint}` for:
- if the project is CNG (without ios/android folders), they should always run eas update from CNG state.
- if the project is bare, they should always run eas update from bare state.
- otherwise, we show an error.

using gitignore to detect the CNG/bare state also makes sense to me.

File: packages/expo-updates/ios/EXUpdates.podspec
Comment (Kudo) [inline packages/expo-updates/ios/EXUpdates.podspec:74]: rather than having a new script phase (and new gradle task), what do you think to rename [createManifest.js](https://github.com/expo/expo/blob/main/packages/expo-updates/scripts/createManifest.js) and **create-manifest-ios.sh** to something like createUpdateResources and merge the manifest / fingerprint invocation entry point all together?

having multiple script phases or gradle tasks may increase the difficulty of debugging when there are any errors from xcode or gradle building system.

**Comment by wschurman** [comment]
> even though expo-updates supports different workflows, we could limit the team running `eas update` from the same workflow. that is exactly what `runtimeVersion={hasPrebuild}:{fingerprint}` for

The issue there is that we don't store the runtime version anywhere in source control (it's a build artifact and generated on the fly in `eas-cli`, and the only thing generated during prebuild is put in the native directories which are ignored or not ignored), so we can't keep track of the "state" of whether the project always uses CNG or has stuff checked in other than detecting whether they're gitignored. (The precise steps this PR does is summarized in the `How` section of the summary).

I updated this PR to:
- keep one policy
- use gitignore(s) to see if the same marker files are ignored that eas-cli uses. note though that we'll need to modify eas-cli to not use `.easignore` for this check to match the check in this library so that a consistent fingerprint is generated (this library shouldn't have any notion of what EAS is if we can help it since it is backend-agnostic and solely uses the spec).
- merge the createManifest and createFingerprint into one build script/step.

File: yarn.lock
Comment (Kudo) [inline yarn.lock:19783]: nit: it happens when running `yarn add` inside packages/expo-updates. please try `rm -rf /path/to/expo/expo/node_modules && yarn` to generate correct yarn.lock.

**Review by Kudo — APPROVED** [review]
thanks for the explanation and updates. it looks good and i like much for generating fingerprint at build time.

**Comment by wschurman** [comment]
Seeing an issue with this where running an xcode release build multiple times in a row doesn't produce a consistent fingerprint. Will need to investigate.

File: packages/@expo/fingerprint/src/Options.ts
Comment (Kudo) [inline packages/@expo/fingerprint/src/Options.ts:16]: thanks for fixing these 👏

**Comment by giorgiofellipe** [comment]
Hi @wschurman.

Am I supposed to see my runtime version as "file:fingerprint" on dev client?
<image src="https://github.com/expo/expo/assets/4589857/302c0089-a3cb-4c8c-90f1-497b90c1d6f2" width="250" />

Also, I can't reach my branches anymore inside extensions:

<image src="https://github.com/expo/expo/assets/4589857/7089b34d-e725-4a2b-b559-9b6083bb02eb" width="250" />

**Comment by wschurman** [comment]
@giorgiofellipe - thanks for the report. Added to list of tasks to complete before moving fingerprint out of beta.

**Comment by giorgiofellipe** [comment]
> @giorgiofellipe - thanks for the report. Added to list of tasks to complete before moving fingerprint out of beta.

You see any workaround I can try while you guys knock it down?

**Comment by wschurman** [comment]
Not that I know of off the top of my head, unfortunately. As the feature is still in beta there are various pieces of the puzzle that still need ironing out.

**Pull request merged**
